packages/mattermost-redux/src/utils/apps.ts

```diff
@@ -1,116 +1,189 @@
 }
 
-// filterInvalidCommands remove commands without a label
-function filterInvalidCommands(binding?: AppBinding) {
-    if (!binding) {
+export function cleanForm(form?: AppForm) {
+    if (!form) {
         return;
     }
 
-    const isValidCommand = (b: AppBinding): boolean => {
-        return Boolean(b.label);
-    };
+    const toRemove: number[] = [];
+    const usedLabels: {[label: string]: boolean} = {};
+    form.fields?.forEach((field, i) => {
+        if (!field.name) {
+            toRemove.unshift(i);
+            return;
+        }
 
-    const validateCommand = (b: AppBinding) => {
-        b.bindings = b.bindings?.filter(isValidCommand);
-        b.bindings?.forEach(validateCommand);
-    };
+        if (field.name.match(/ |\t/)) {
+            toRemove.unshift(i);
+            return;
+        }
 
-    binding.bindings?.filter((b) => b.location === AppBindingLocations.COMMAND).forEach(validateCommand);
-}
+        let label = field.label;
+        if (!label) {
+            label = field.name;
+        }
 
-// filterInvalidChannelHeaderBindings remove bindings
-// without a label or without an icon.
-function filterInvalidChannelHeaderBindings(binding?: AppBinding) {
-    if (!binding) {
-        return;
-    }
+        if (label.match(/ |\t/)) {
+            toRemove.unshift(i);
+            return;
+        }
+
+        if (usedLabels[label]) {
+            toRemove.unshift(i);
+            return;
+        }
 
-    const isValidChannelHeaderBindings = (b: AppBinding): boolean => {
-        return Boolean(b.icon && b.label);
-    };
+        if (field.type === AppFieldTypes.STATIC_SELECT) {
+            cleanStaticSelect(field);
+            if (!field.options?.length) {
+                toRemove.unshift(i);
+                return;
+            }
+        }
 
-    const validateChannelHeaderBinding = (b: AppBinding) => {
-        b.bindings = b.bindings?.filter(isValidChannelHeaderBindings);
-        b.bindings?.forEach(validateChannelHeaderBinding);
-    };
+        usedLabels[label] = true;
+    });
 
-    binding.bindings?.filter((b) => b.location === AppBindingLocations.CHANNEL_HEADER_ICON).forEach(validateChannelHeaderBinding);
+    toRemove.forEach((i) => {
+        form.fields.splice(i, 1);
+    });
 }
 
-// filterInvalidPostMenuBindings remove bindings
-// without a label.
-function filterInvalidPostMenuBindings(binding?: AppBinding) {
-    if (!binding) {
-        return;
-    }
+function cleanStaticSelect(field: AppField) {
+    const toRemove: number[] = [];
+    const usedLabels: {[label: string]: boolean} = {};
+    const usedValues: {[label: string]: boolean} = {};
+    field.options?.forEach((option, i) => {
+        let label = option.label;
+        if (!label) {
+            label = option.value;
+        }
+
+        if (!label) {
+            toRemove.unshift(i);
+            return;
+        }
```

Comment:
```suggestion
        if (!label) {
            label = option.value;
        }

        if (!label) {
            toRemove.unshift(i);
            return;
        }
```
could you combine these two?

Reply:
It is weird to combine. We could nest the if, but still would feel weird:
```
if (!label) {
    if (!option.value) {
        toRemove.unshift(i);
        return;
    }
    label = option.value;
}
```
or
```
if (!label) {
    label = option.value;
    if (!label) {
        toRemove.unshift(i);
        return;
    }
}
```
We could turn it around, but it would be pretty similar still:
```
if (!label && !option.value) {
    toRemove.unshift(i);
    return;
}
if (!label) {
    label = option.value;
}
```

Reply:
That explanation makes sense.  When I glanced at this code I didn't account for `option.value` possibly being undefined.